### ISSUE TITLE
Make it possible to use a custom SnapPrevention

### ISSUE THREAD

**Issue by boldtrn** [graphhopper/graphhopper#1713]
The snap prevention feature is great. I just wanted to modify the snap prevention in a project that uses GraphHopper as a library. Unfortunately, it's not possible to override the snap prevention easily. It would be nice to have an easy way to do this.

One thing that could already help would be to move the RoutingTemplate creation in the GraphHopper class into an overridable method:
```
// Move this part into a method
            RoutingTemplate routingTemplate;
            if (ROUND_TRIP.equalsIgnoreCase(algoStr))
                routingTemplate = new RoundTripRoutingTemplate(request, ghRsp, locationIndex, encodingManager, maxRoundTripRetries);
            else if (ALT_ROUTE.equalsIgnoreCase(algoStr))
                routingTemplate = new AlternativeRoutingTemplate(request, ghRsp, locationIndex, encodingManager);
            else
                routingTemplate = new ViaRoutingTemplate(request, ghRsp, locationIndex, encodingManager);
```

Then we could allow passing different EdgeFilters in the constructor of the RoutingTemplate or make the ViaRoutingTemplate extendable to use different EdgeFilters.

WDYT?

**Comment by karussell** [comment]
What exactly do you have in mind that does not make sense in the general case?

**Comment by boldtrn** [comment]
I guess there are many specific examples that are not easy to cover by the general case.

Here are some examples:
- I might want to create a new EncodedValue "snapable" and it should only snap if snapable = X. 
- I don't want to snap edges with `motorway & tunnel || motorway & bridge`. In general "more complex" scenarios than a single value.
- I only want the snap prevention to happen for the start and end, but via points are not considered.

This issue is a mix of allow to pass a custom `RoutingTemplate` and customize `SnapPrevention`. There have been a couple of cases in the past where I'd wanted to customize the RoutingTemplate and was able to work around the issue some other way, by changing the Graph or a different component of GraphHopper to get the same result. Maybe I should split this issue in two issues?

**Comment by karussell** [comment]
Can you look into the fix for #729 if this fixes your use case too? I.e. using a custom Weighting.

**Comment by boldtrn** [comment]
Thanks for the update @karussell. Technically, this would mostly solve this issue, yes.

It would require some special logic like:
```java
    @Override
    public double calcWeight(EdgeIteratorState edgeState, boolean reverse, int prevOrNextEdgeId) {
        if (prevOrNextEdgeId == NO_EDGE){
           // special handling for first edge
        }
        // regular weight handling
    }
``` 

The only thing that would probably not work is to only use the snap prevention for the start and end, and ignore via points.
